### PR TITLE
24147 Create Event Grid domain topic in response to D365 webhook background process

### PR DESCRIPTION
**Pull request by Shirin14926**
Created Web Job project.
Add message to Queue
Coding related to create Event Domain Topic.
Unit test for the same.
DevOps Changes for Event Grid Domain Topic, Storage

### CODE REVIEW

**Review by mattwin1 — CHANGES_REQUESTED** [review]
Minor changes needed: TLS 1>1.2 for storage account, and appsettings.json files need to remove "dev" values from EventHubLoggingConfiguration.Environment vars

**Comment by tejasi15072** [comment]
> Minor changes needed: TLS 1>1.2 for storage account, and appsettings.json files need to remove "dev" values from EventHubLoggingConfiguration.Environment vars

-Requested changes are implemented.

**Review by usha13529 — CHANGES_REQUESTED** [review]
Please fix review comments.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Services/ISubscriptionService.cs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ﻿using System.Threading.Tasks;
 using FluentValidation.Results;
+using UKHO.ExternalNotificationService.Common.Configuration;
 using UKHO.ExternalNotificationService.Common.Models.Request;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Services/ISubscriptionService.cs:5]: Did short

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@
 using UKHO.ExternalNotificationService.API.Validation;
 using UKHO.ExternalNotificationService.Common.Configuration;
 using UKHO.ExternalNotificationService.Common.HealthCheck;
+using UKHO.ExternalNotificationService.Common.Helpers;
 using UKHO.ExternalNotificationService.Common.Repository;
 using UKHO.Logging.EventHubLogProvider;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs:25]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/SubscriptionStorageConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/SubscriptionStorageConfiguration.cs]: yes removed

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            _eventGridDomainConfig = eventGridDomainConfig.Value;
+            _logger = logger;            
+        }
+        
+        public async Task<string> CreateOrUpdateSubscription(SubscriptionRequestMessage subscriptionRequestMessage, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation(EventIds.CreateOrUpdateAzureEventDomainTopicStart.ToEventId(),
+                    "Create or update azure event domain topic started for _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId} with Event domain topic {NotificationTypeTopicName}", subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId, subscriptionRequestMessage.NotificationTypeTopicName);
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs:33]: added subscription id in log

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            };
+            
+            QueueClient queueClient = new(storageAccountConnectionString, ensStorageConfiguration.QueueName, queueClientOptions);
+
+            string subscriptionMessageString = JsonSerializer.Serialize(subscriptionRequestMessage);
+            // Send a message to the queue
+            await queueClient.SendMessageAsync(subscriptionMessageString);           
+            
+            _logger.LogInformation(EventIds.AddedMessageInQueue.ToEventId(), "Added message in Queue for message:{subscriptionMessageString} and _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{correlationId}", subscriptionMessageString, subscriptionRequestMessage , correlationId);
+        }
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs:37]: added subscription id in log

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Models/Request/SubscriptionRequest.cs [hunk 6]
```diff
@@ -1,5 +1,4 @@
-﻿
-namespace UKHO.ExternalNotificationService.Common.Models.Request
+﻿namespace UKHO.ExternalNotificationService.Common.Models.Request
 {
     public class SubscriptionRequest
     {
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Models/Request/SubscriptionRequest.cs:4]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Serilog;
+using Serilog.Events;
+using UKHO.ExternalNotificationService.Common.Configuration;
+using UKHO.ExternalNotificationService.Common.Helpers;
+using UKHO.ExternalNotificationService.SubscriptionService.Services;
+using UKHO.Logging.EventHubLogProvider;
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs:17]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+namespace UKHO.ExternalNotificationService.API.FunctionalTests.FunctionalTests
+{
+    class EnsSubscriptionRequestInQueueTest
+    {
+        private EnsApiClient _ensApiClient { get; set; }
+        private TestConfiguration _testConfig { get; set; }
+        private D365Payload _d365Payload { get; set; }
+        private QueueClient _queue { get; set; }
+        private EventGridManagementClient _eventGridMgmtClient { get; set; }
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs:25]: did changes accordingly

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+using System.Threading.Tasks;
+using Azure.Storage.Queues;
+using Azure.Storage.Queues.Models;
+using Microsoft.Azure.Management.EventGrid;
+using Microsoft.Azure.Management.EventGrid.Models;
+using Newtonsoft.Json;
+using NUnit.Framework;
+using UKHO.ExternalNotificationService.API.FunctionalTests.Helper;
+using UKHO.ExternalNotificationService.API.FunctionalTests.Model;
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs:15]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.csproj [hunk 10]
```diff
@@ -4,7 +4,14 @@
...
   </PropertyGroup>
 
-   <ItemGroup>
+  <ItemGroup>
+    <Compile Remove="Configuration\**" />
+    <Content Remove="Configuration\**" />
+    <EmbeddedResource Remove="Configuration\**" />
+    <None Remove="Configuration\**" />
+  </ItemGroup>
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.csproj]: Not sure but as of now it is removed. Might be it had added with any changes related to Nuget etc. But yse now i have removed this.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            _logger.LogInformation(EventIds.CreateOrUpdateAzureEventDomainTopicStart.ToEventId(),
+                    "Create or update azure event domain topic started for _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId} with Event domain topic {NotificationTypeTopicName}", subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId, subscriptionRequestMessage.NotificationTypeTopicName);
+            
+            EventGridManagementClient eventGridMgmtClient = await GetEventGridClient(_eventGridDomainConfig.SubscriptionId, cancellationToken);           
+            DomainTopic topic = await GetDomainTopic(eventGridMgmtClient, subscriptionRequestMessage.NotificationTypeTopicName, cancellationToken);
+
+            _logger.LogInformation(EventIds.CreateOrUpdateAzureEventDomainTopicCompleted.ToEventId(),
+                    "Create or update azure event domain topic completed for _D365-Correlation-ID:{correlationId} and _X-Correlation-ID:{CorrelationId} with Event domain topic {topic}", subscriptionRequestMessage.D365CorrelationId, subscriptionRequestMessage.CorrelationId, topic.Name);
+            return topic.Id;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs:39]: added

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/EventGridDomainConfiguration.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/EventGridDomainConfiguration.cs]: done

**Review by usha13529 — APPROVED** [review]
LGTM

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+namespace UKHO.ExternalNotificationService.API.FunctionalTests.FunctionalTests
+{
+    class EnsSubscriptionRequestInQueueTest
+    {
+        private EnsApiClient _ensApiClient { get; set; }
+        private TestConfiguration _testConfig { get; set; }
+        private D365Payload _d365Payload { get; set; }
+        private QueueClient _queue { get; set; }
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs:20]: Correct naming of properties please:
```suggestion
        private EnsApiClient EnsApiClient { get; set; }
        private TestConfiguration TestConfig { get; set; }
        private D365Payload D365Payload { get; set; }
        private QueueClient Queue { get; set; }        
```

Do you not get these suggestions in VS?:
image

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            {
+                host.Run();
+            }
+        }
+
+        private static HostBuilder BuildHostConfiguration()
+        {
+
+            HostBuilder hostBuilder = new();
+            hostBuilder.ConfigureAppConfiguration((hostContext, builder) =>
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+                IConfigurationRoot tempConfig = builder.Build();
+                string kvServiceUri = tempConfig["KeyVaultSettings:ServiceUri"];
+                if (!string.IsNullOrWhiteSpace(kvServiceUri))
+                {
+                    builder.AddAzureKeyVault(new Uri(kvServiceUri), new DefaultAzureCredential());
+                }
+
+                Program.ConfigurationBuilder = builder.Build();
+            })
+             .ConfigureLogging((hostContext, builder) =>
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+using UKHO.ExternalNotificationService.SubscriptionService.Services;
+using UKHO.Logging.EventHubLogProvider;
+
+namespace UKHO.ExternalNotificationService.SubscriptionService
+{
+    [ExcludeFromCodeCoverage]
+    public static class Program
+    {
+        private static IConfiguration ConfigurationBuilder;
+        private static string AssemblyVersion = Assembly.GetExecutingAssembly().GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+                         config.NodeName = eventhubConfig.NodeName;
+                         config.AdditionalValuesProvider = additionalValues =>
+                         {
+                             additionalValues["_AssemblyVersion"] = AssemblyVersion;
+                         };
+                     });
+                 }
+
+             })
+              .ConfigureServices((hostContext, services) =>
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService/Program.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/SubscriptionServiceDataTests.cs [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        private ILogger<SubscriptionServiceData> _fakeLogger;
+        private SubscriptionServiceData _fakeSubscriptionServiceData;
+
+        [SetUp]
+        public void Setup()
+        {
+            _fakeAzureEventGridDomainService = A.Fake<IAzureEventGridDomainService>();
+            _fakeLogger = A.Fake<ILogger<SubscriptionServiceData>>();
+
+            _fakeSubscriptionServiceData = new SubscriptionServiceData(_fakeAzureEventGridDomainService, _fakeLogger);
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/SubscriptionServiceDataTests.cs]: yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Models/Request/SubscriptionRequestMessage.cs [hunk 19]
```diff
@@ -0,0 +1,19 @@
+﻿namespace UKHO.ExternalNotificationService.Common.Models.Request
+{
+    public class SubscriptionRequestMessage
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Models/Request/SubscriptionRequestMessage.cs]: yes moved to base class

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Threading.Tasks;
+using UKHO.ExternalNotificationService.Common.Configuration;
+using UKHO.ExternalNotificationService.Common.Logging;
+using UKHO.ExternalNotificationService.Common.Models.Request;
+
+namespace UKHO.ExternalNotificationService.Common.Helpers
+{
+    [ExcludeFromCodeCoverage]
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs:12]: As we are using QueueClient which is used to create actual connection to Azure Queue and want be able to mock that. So, it is been excluded and as in ESS we have excluded too. Thanks!

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+            EventGridManagementClient _egClient = new(credential)
+            {
+                SubscriptionId = SubscriptionId
+            };
+
+            return _egClient;
+        }
+
+        protected virtual async Task<DomainTopic> GetDomainTopic(EventGridManagementClient eventGridMgmtClient, string NotificationTypeTopicName, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs [hunk 22]
```diff
@@ -109,9 +112,11 @@ public async Task WhenPostValidPayload_ThenReceiveSuccessfulResponse()
             A.CallTo(() => _fakeSubscriptionService.ValidateD365PayloadRequest(A<D365Payload>.Ignored)).Returns(new ValidationResult(new List<ValidationFailure>()));
             A.CallTo(() => _fakeSubscriptionService.ConvertToSubscriptionRequestModel(A<D365Payload>.Ignored)).Returns(_fakeSubscriptionRequest);
             A.CallTo(() => _fakeNotificationRepository.GetAllNotificationTypes()).Returns(_fakeNotificationType);
+
             var result = (StatusCodeResult)await _controller.Post(_fakeD365PayloadDetails);
 
             A.CallTo(_fakeLogger).Where(call => call.GetArgument<LogLevel>(0) == LogLevel.Error).MustNotHaveHappened();
+            A.CallTo(() => _fakeSubscriptionService.AddSubscriptionRequest(A<SubscriptionRequest>.Ignored, A<NotificationType>.Ignored, A<string>.Ignored)).MustHaveHappenedOnceExactly();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+            Assert.AreEqual(202, (int)apiResponse.StatusCode, $"Incorrect status code {apiResponse.StatusCode}  is  returned, instead of the expected 202.");
+
+            QueueMessage[] messageQueue = _queue.ReceiveMessages();
+
+            DateTime startTime = DateTime.UtcNow;
+
+            while (messageQueue.Length < 1 && DateTime.UtcNow - startTime < TimeSpan.FromSeconds(_testConfig.WaitingTimeForQueueInSeconds))
+            {
+                await Task.Delay(1000);
+                messageQueue = _queue.ReceiveMessages();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+            
+        }
+        
+        [Test]
+        public async Task WhenICallTheEnsSubscriptionApiWithAValidD365Payload_ThenMessageAddedInQueue()
+        {
+            HttpResponseMessage apiResponse = await _ensApiClient.PostEnsApiSubscriptionAsync(_d365Payload);
+            Assert.AreEqual(202, (int)apiResponse.StatusCode, $"Incorrect status code {apiResponse.StatusCode}  is  returned, instead of the expected 202.");
+
+            QueueMessage[] messageQueue = _queue.ReceiveMessages();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs (hunk 13 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.FunctionalTests/FunctionalTests/EnsSubscriptionRequestInQueueTest.cs]: Thanks! done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs [hunk 25]
```diff
@@ -19,8 +21,12 @@ public class SubscriptionServiceTest
         private ID365PayloadValidator _fakeD365PayloadValidator;
         private IOptions<D365PayloadKeyConfiguration> _fakeD365PayloadKeyConfiguration;
         private D365Payload _fakeD365PayloadDetails;
-        private SubscriptionRequest _fakeSubscriptionRequest;
-        private SubscriptionService _subscriptionService;
+        private SubscriptionRequest _fakeSubscriptionRequest;        
+        private NotificationType _fakeNotificationType;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs]: done changes accordingly

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs [hunk 26]
```diff
@@ -114,6 +127,17 @@ public void WhenValidPayloadInRequest_ThenReceiveSubscriptionRequest()
         }
         #endregion
 
+        [Test]
+        public void WhenValidSubscriptionRequestDetailsPassed_ThenAddMessageInQueue()
+        {
+            string fakeCorrelationid = "6ea03f10-2672-46fb-92a1-5200f6a4faaa";
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs [hunk 27]
```diff
@@ -114,6 +127,17 @@ public void WhenValidPayloadInRequest_ThenReceiveSubscriptionRequest()
...
         #endregion
 
+        [Test]
+        public void WhenValidSubscriptionRequestDetailsPassed_ThenAddMessageInQueue()
+        {
+            string fakeCorrelationid = "6ea03f10-2672-46fb-92a1-5200f6a4faaa";            
+            
+            var result = _fakeSubscriptionService.AddSubscriptionRequest(_fakeSubscriptionRequest, _fakeNotificationType, fakeCorrelationid);
+            
+            A.CallTo(() => _fakeAzureMessageQueueHelper.AddQueueMessage(A<SubscriptionStorageConfiguration>.Ignored, A<SubscriptionRequestMessage>.Ignored)).MustHaveHappenedOnceExactly();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs]: as **SubscriptionRequestMessage** object is been created and used with in the **AddSubscriptionRequest** method. So, could not use it for assert. Rest of the thing is done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs [hunk 28]
```diff
@@ -19,8 +21,12 @@ public class SubscriptionServiceTest
         private ID365PayloadValidator _fakeD365PayloadValidator;
         private IOptions<D365PayloadKeyConfiguration> _fakeD365PayloadKeyConfiguration;
         private D365Payload _fakeD365PayloadDetails;
-        private SubscriptionRequest _fakeSubscriptionRequest;
-        private SubscriptionService _subscriptionService;
+        private SubscriptionRequest _fakeSubscriptionRequest;        
+        private NotificationType _fakeNotificationType;
+        private SubscriptionService _fakeSubscriptionService;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/SubscriptionServiceDataTests.cs [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        [Test]
+        public async Task WhenCreateOrUpdateSubscriptionThenCreateSubscription()
+        {
+            CancellationToken cancellationToken = CancellationToken.None;
+
+            string response = await _fakeSubscriptionServiceData.CreateOrUpdateSubscription(GetSubscriptionRequestMessage(), cancellationToken);
+
+            A.CallTo(() => _fakeAzureEventGridDomainService.CreateOrUpdateSubscription(A<SubscriptionRequestMessage>.Ignored, A<CancellationToken>.Ignored)).MustHaveHappenedOnceExactly();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/SubscriptionServiceDataTests.cs]: done

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+        private static async Task<EventGridManagementClient> GetEventGridClient(string SubscriptionId, CancellationToken cancellationToken)
+        {
+            DefaultAzureCredential azureCredential = new();
+            TokenRequestContext tokenRequestContext = new(new string[] { "https://management.azure.com/.default" });
+
+            AccessToken tokenResult = await azureCredential.GetTokenAsync(tokenRequestContext, cancellationToken);
+            TokenCredentials credential = new(tokenResult.Token);
+
+            EventGridManagementClient _egClient = new(credential)
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureEventGridDomainService.cs]: Now not using any variable. used return new(....){.......}

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (barrie-cooper) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Helpers/AzureMessageQueueHelper.cs:12]: Dammit, you're right 😁

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs [hunk 31]
```diff
@@ -18,9 +20,13 @@ public class SubscriptionServiceTest
     {
         private ID365PayloadValidator _fakeD365PayloadValidator;
         private IOptions<D365PayloadKeyConfiguration> _fakeD365PayloadKeyConfiguration;
-        private D365Payload _fakeD365PayloadDetails;
-        private SubscriptionRequest _fakeSubscriptionRequest;
-        private SubscriptionService _subscriptionService;
+        private D365Payload _d365PayloadDetails;
+        private SubscriptionRequest _subscriptionRequest;        
+        private NotificationType _notificationType;
+        private ISubscriptionService subscriptionService;
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs]: Naming please:
```suggestion
        private ISubscriptionService _subscriptionService;
```

**Review by barrie-cooper — COMMENTED** [review]
Just two small naming issues

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/SubscriptionServiceDataTests.cs [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+using UKHO.ExternalNotificationService.SubscriptionService.Services;
+
+namespace UKHO.ExternalNotificationService.Webjob.UnitTests.Services
+{
+    [TestFixture]
+    public class SubscriptionServiceDataTests
+    {
+        private IAzureEventGridDomainService _fakeAzureEventGridDomainService;
+        private ILogger<SubscriptionServiceData> _fakeLogger;
+        private SubscriptionServiceData subscriptionServiceData;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.SubscriptionService.UnitTests/Services/SubscriptionServiceDataTests.cs:17]: That's done while merging main.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs (hunk 31 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Services/SubscriptionServiceTest.cs]: that's done while merging min to this branch

**Review by barrie-cooper — APPROVED** [review]
Thanks for changes

**Pull request merged**